src/careamics/dataset_ng/patch_extractor/image_stack_loader.py

```diff
@@ -0,0 +1,138 @@
+    Optional,
+    Protocol,
+    Union,
+)
+
+from numpy.typing import NDArray
+from typing_extensions import ParamSpec
+
+from careamics.config import DataConfig
+from careamics.config.support import SupportedData
+from careamics.file_io.read import ReadFunc
+
+from .image_stack import ImageStack, InMemoryImageStack, ZarrImageStack
+
+P = ParamSpec("P")
+
+
+class ImageStackLoader(Protocol[P]):
+    """
+    Protocol to define how `ImageStacks` should be loaded.
+
+    An `ImageStackLoader` is a callable that must take the CAREamics `DataConfig` as
+    the first argument and the `source` of the data as the second argument.
+    Additional `*args` and `**kwargs` are allowed, but they should only be used to
+    determine _how_ the data is loaded, not _what_ data is loaded. The `source`
+    argument has to wholly determine _what_ data is loaded, this is because,
+    downstream, both an input-source and a target-source have to be specified but they
+    will share `*args` and `**kwargs`.
+
+    An `ImageStackLoader` must return a sequence of the `ImageStack` class. This could
+    be a sequence of one of the existing concrete implementations, such as
+    `ZarrImageStack`, or a custom user defined `ImageStack`.
+
+    Example
+    -------
+    The following example demonstrates how an `ImageStackLoader` could be defined
+    for loading non-OME Zarr images. Returning a list of `ZarrImageStack` instances.
+
+    >>> from typing import TypedDict
+
+    >>> from zarr.storage import FSStore
+
+    >>> from careamics.config import DataConfig
+    >>> from careamics.dataset_ng.patch_extractor.image_stack import ZarrImageStack
+
+    >>> # Define a zarr source
+    >>> # It encompasses multiple arguments that determine what data will be loaded
+    >>> class ZarrSource(TypedDict):
+    ...     store: FSStore
+    ...     data_paths: Sequence[str]
+
+    >>> def custom_image_stack_loader(
+    ...     data_config: DataConfig, source: ZarrSource, *args, **kwargs
+    ... ) -> list[ZarrImageStack]:
+    ...     axes = data_config.axes
+    ...     image_stacks = [
+    ...         ZarrImageStack(store=source["store"], data_path=data_path, axes=axes)
+    ...         for data_path in source["data_paths"]
+    ...     ]
+    ...     return image_stacks
+
+    TODO: show example use in the `CAREamicsDataset`
+
+    The example above defines a `ZarrSource` dict because to determine _which_ ZARR
+    images will be loaded both a ZARR store and the internal data paths need to be
+    specified.
+    """
+
+    def __call__(
+        self, data_config: DataConfig, source: Any, *args: P.args, **kwargs: P.kwargs
+    ) -> Sequence[ImageStack]: ...
+
+
+def from_arrays(
+    data_config: DataConfig, source: Sequence[NDArray], *args, **kwargs
+) -> list[InMemoryImageStack]:
+    axes = data_config.axes
+    return [InMemoryImageStack.from_array(data=array, axes=axes) for array in source]
+
+
+# TODO: change source to directory path? Like in current implementation
```

Comment:
Both could be nice to have. Leaving users in charge expose them to an error prone process. In my experience, people like being able to place everyhting in a folder with matching names between two folders, so it is a desirable feature.